src/platforms/atomic-kms/server/kms/atomic_kms_output.cpp

```diff
@@ -0,0 +1,911 @@
+    {
+        mir::log_error("Output %s has no associated CRTC to set a framebuffer on",
+                       mgk::connector_name(conf->connector).c_str());
+        return false;
+    }
+
+    /* We use the *requested* mode rather than the current_crtc
+     * because we might have been asked to perform a modeset
+     */
+    auto const width = conf->connector->modes[conf->mode_index].hdisplay;
+    auto const height = conf->connector->modes[conf->mode_index].vdisplay;
+
+    AtomicUpdate update;
+    update.add_property(*conf->crtc_props, "MODE_ID", conf->mode->handle());
+    update.add_property(*conf->connector_props, "CRTC_ID", conf->current_crtc->crtc_id);
+    update.add_property(*conf->crtc_props, "ACTIVE", 1);
+
+    /* Source viewport. Coordinates are 16.16 fixed point format */
+    update.add_property(*conf->plane_props, "SRC_X", conf->fb_offset.dx.as_uint32_t() << 16);
+    update.add_property(*conf->plane_props, "SRC_Y", conf->fb_offset.dy.as_uint32_t() << 16);
+    update.add_property(*conf->plane_props, "SRC_W", width << 16);
+    update.add_property(*conf->plane_props, "SRC_H", height << 16);
+
+    /* Destination viewport. Coordinates are *not* 16.16 */
+    update.add_property(*conf->plane_props, "CRTC_X", 0);
+    update.add_property(*conf->plane_props, "CRTC_Y", 0);
+    update.add_property(*conf->plane_props, "CRTC_W", width);
+    update.add_property(*conf->plane_props, "CRTC_H", height);
+
+    /* Set a surface for the plane */
+    update.add_property(*conf->plane_props, "CRTC_ID", conf->current_crtc->crtc_id);
+    update.add_property(*conf->plane_props, "FB_ID", fb);
+
+    auto ret = drmModeAtomicCommit(drm_fd_, update, DRM_MODE_ATOMIC_ALLOW_MODESET, nullptr);
+    if (ret)
+    {
+        mir::log_error("Failed to set CRTC: %s (%i)", strerror(-ret), -ret);
+        conf->current_crtc = nullptr;
+        return false;
+    }
+
+    // We might have performed a modeset; update our view of the hardware state accordingly!
+    conf->current_crtc = mgk::get_crtc(drm_fd_, conf->current_crtc->crtc_id);
+
+    using_saved_crtc = false;
+    return true;
+}
+
+bool mga::AtomicKMSOutput::has_crtc_mismatch()
+{
+    auto conf = configuration.lock();
+    if (!ensure_crtc(*conf))
+    {
+        mir::log_error("Output %s has no associated CRTC to get ", mgk::connector_name(conf->connector).c_str());
+        return true;
+    }
+
+    return !kms_modes_are_equal(&conf->current_crtc->mode, &conf->connector->modes[conf->mode_index]);
+}
+
+void mga::AtomicKMSOutput::clear_crtc()
+{
+    auto conf = configuration.lock();
+    try
+    {
+        ensure_crtc(*conf);
+    }
+    catch (...)
+    {
+        /*
+         * In order to actually clear the output, we need to have a crtc
+         * connected to the output/connector so that we can disconnect
+         * it. However, not being able to get a crtc is OK, since it means
+         * that the output cannot be displaying anything anyway.
+         */
+        return;
+    }
```

Comment:
A try block around a single line of code followed by a catch block that silently eats everything suggests we are using the wrong reporting mechanism.